tests/system_tests.py

```diff
@@ -406,42 +408,75 @@ def setUp_file_action(self, expanded_file_name):
         return shutil.copyfile(expanded_file_name, new_name)
 
 
+def test_run(self):
+    """
+    This function reads in the members commands, retval, stdout, stderr and runs
+    the `expand_variables` function on each. The resulting commands are then run
+    using the subprocess module and compared against the expected values that
+    were provided in the static members via `compare_stdout` and
+    `compare_stderr`. Furthermore a threading.Timer is used to abort the
+    execution if a configured timeout is reached.
+
+    It is automatically added as a member function to each system test by the
+    CaseMeta metaclass. This ensures that it is run by each system test
+    **after** setUp() and setUpClass() were run.
+    """
+    for i, command, retval, stdout, stderr in zip(range(len(self.commands)),
+                                                  self.commands,
+                                                  self.retval,
+                                                  self.stdout,
+                                                  self.stderr):
+        command, retval, stdout, stderr = map(
+            self.expand_variables, [command, retval, stdout, stderr]
+        )
+        retval = int(retval)
+        timeout = {"flag": False}
+
+        proc = subprocess.Popen(
+            _cmd_splitter(command),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            cwd=self.work_dir
+        )
+
+        def timeout_reached(timeout):
+            timeout["flag"] = True
+            proc.kill()
+
+        t = threading.Timer(
+            _parameters["timeout"], timeout_reached, args=[timeout]
+        )
+        t.start()
+        got_stdout, got_stderr = proc.communicate()
+        t.cancel()
+
+        processed_stdout = _process_output_post(got_stdout.decode('utf-8'))
+        processed_stderr = _process_output_post(got_stderr.decode('utf-8'))
+
+        self.assertFalse(timeout["flag"] and "Timeout reached")
```

Comment:
Did you want to show here the error message "Timeout reached" in case the timeout is True? I think, right now the string is part of the assert.

Reply:
Yes, I did it the C-way. I have just pushed another commit and added the message as the `msg` parameter to the `assert` (I forgot about the `msg` parameter when I wrote the first version of the `test_run` function).